File: week1/2.add_two_nums-链表两数相加/2.add_two_nums.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
 var addTwoNumbers = function(l1, l2) {
     let pre = new ListNode();
     let cur = pre;
     let carry = 0;

     while (l1 != null || l2 != null) {
         let x = l1 == null ? 0 : l1.val;
         let y = l2 == null ? 0 : l2.val;
         let sum = x + y + carry;
        
         carry = Math.floor(sum / 10);
         sum = sum % 10;
         cur.next = new ListNode(sum);
         cur = cur.next;

         if (l1 != null) {
             l1 = l1.next;
         }
         if (l2 != null) {
             l2 = l2.next;
         }
     }
     if (carry > 0 ) {
         cur.next = new ListNode(carry);
     }

     return pre.next;
};

File: Algorithms/0002.add_two_nums-链表两数相加/2.add_two_nums.go
package problem0002

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1, l2 *ListNode) (head *ListNode) {
	var tail *ListNode
	carry := 0
	for l1 != nil || l2 != nil {
		n1, n2 := 0, 0
		if l1 != nil {
			n1 = l1.Val
			l1 = l1.Next
		}

		if l2 != nil {
			n2 = l2.Val
			l2 = l2.Next
		}

		sum := n1 + n2 + carry
		sum, carry = sum%10, sum/10

		if head == nil {
			head = &ListNode{Val: sum}
			tail = head
		} else {
			tail.Next = &ListNode{Val: sum}
			tail = tail.Next
		}

	}
	if carry > 0 {
		tail.Next = &ListNode{Val: carry}
	}
	return
}

func addTwo(l1, l2 *ListNode) *ListNode {
	head := &ListNode{}
	cur := head
	carry := 0

	for l1 != nil || l2 != nil || carry > 0 {
		sum := carry

		if l1 != nil {
			sum += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			sum += l2.Val
			l2 = l2.Next
		}

		carry = sum / 10
		cur.Next = &ListNode{Val: sum % 10}
		cur = cur.Next
	}
	return head.Next
}

/*
	test case
	Input :
	[2,4,3]
	[5,6,4]

	Output:
	[7,0,8]
*/
